### PR TITLE
Implement ADD_RESPONSE action

### LINKED ISSUE #19: TP: Implement ADD_RESPONSE action
Include tests to demonstrate it works. Blocked by #17.

## Original Design

Adds a response to an Offer.

### Requirements

- Signer must be registered as a collection
- Signer must own the cryptomoji or be the owner of the Offer
- All cryptomoji must exist 

### Payload Schema

```
{
  "action": "ADD_RESPONSE",
  "offer": <string, address>,
  "moji": [ <strings, addresses> ]
}
```
### State Schema

```
{
  "owner": <string, public key>,
  "moji": [ <strings, addresses> ],
  "responses": [
    {
      "owner": <string, public key>,
      "moji": [ <strings, addresses> ],
      "isRequest": <bool>
    }
  ]
}
```

### PR DESCRIPTION
**Pull request by delventhalz**
Based on #50 and #53, only the last three commits are a part of this PR.

Adds tests for and implements the ADD_RESPONSE action. Tests can be be run with `npm test` from the processor directory.

Closes #19

### CODE REVIEW

File: code/part-two/processor/test/06-AddResponse.js [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+const { expect } = require('chai');
+const { InvalidTransaction } = require('sawtooth-sdk/processor/exceptions');
+
+const MojiHandler = require('../handler');
+const getAddress = require('../utils/addressing');
+const { decode } = require('../utils/helpers');
+const Txn = require('./mocks/txn');
+const Context = require('./mocks/context');
+
+describe('Create Offer', function() {
```
Comment (therobinkim) [inline code/part-two/processor/test/06-AddResponse.js]: should be `'Add Response'` instead?

File: code/part-two/processor/test/06-AddResponse.js [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+
+    return handler.apply(txn, context)
+      .then(() => {
+        const { responses } = decode(context._state[offer]);
+        expect(responses[0].approver, 'Response should need responder approval')
+          .to.equal(responderPub);
+      });
+  });
+
+  it('should sort the moji in the response', function() {
```
Comment (therobinkim) [inline code/part-two/processor/test/06-AddResponse.js]: is the ln 94 test of `should sort the moji in the response` necessary if we have the ln 106 test of `should reject responses identical to one previously added`?

File: code/part-two/processor/actions/add_response.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+      if (decodedOffer.owner === decodedMoji[0].owner) {
+        const message = 'Response moji and offer have same owner:';
+        return reject(message, decodedOffer.owner);
+      }
+
+      const approver = signerIsOwner
+        ? decodedMoji[0].owner
+        : decodedOffer.owner;
+
+      for (let offered of decodedMoji) {
```
Comment (therobinkim) [inline code/part-two/processor/actions/add_response.js]: possible to not use the word `offer` here? my brain is falling apart trying to remember that `let offered` is actually each moji that is responded with

maybe `let responseItem`? 🤔 

(lmk if I'm just entirely misinterpreting/missing the point!)

**Comment by therobinkim** [comment]
PR not approved on my end yet; waiting on responses to discussion items above.

File: code/part-two/processor/test/06-AddResponse.js (hunk 2 above)
Comment (delventhalz) [reply code/part-two/processor/test/06-AddResponse.js]: Yeah, I think we need it. You could check if pass the test on 106 without sorting the moji you store. I will add this test to CREATE_OFFER as well.

File: code/part-two/processor/actions/add_response.js (hunk 3 above)
Comment (delventhalz) [reply code/part-two/processor/actions/add_response.js]: Changed to `moji`, and removed the other uses of `offered` as well.

File: code/part-two/processor/actions/add_response.js [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        }
+
+        if (!signerIsOwner && moji.owner !== publicKey) {
+          return reject('Signer must own specified moji:', moji.address);
+        }
+      }
+
+      const listing = getAddress.sireListing(decodedMoji[0].owner);
+
+
```
Comment (therobinkim) [inline code/part-two/processor/actions/add_response.js]: extra newlines on lns 71-72 and lns 75-76

**Pull request merged**